awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Visualizer.jsx

```diff
@@ -1,94 +1,688 @@
-import React, { useState, useEffect } from 'react';
+import React, { Fragment, useState, useEffect } from 'react';
+import { withRouter } from 'react-router-dom';
 import { withI18n } from '@lingui/react';
 import { t } from '@lingui/macro';
 import styled from 'styled-components';
+import { shape } from 'prop-types';
+import { BaseSizes, Title, TitleLevel } from '@patternfly/react-core';
+import { layoutGraph } from '@components/Workflow/WorkflowUtils';
 import ContentError from '@components/ContentError';
 import ContentLoading from '@components/ContentLoading';
-import Graph from './Graph';
-import StartScreen from './StartScreen';
-import Toolbar from './Toolbar';
-import { WorkflowJobTemplatesAPI } from '@api';
+import {
+  DeleteAllNodesModal,
+  LinkModal,
+  LinkDeleteModal,
+  NodeDeleteModal,
+  NodeViewModal,
+  UnsavedChangesModal,
+} from './Modals';
+import { NodeModal } from './Modals/NodeModal';
+import VisualizerGraph from './VisualizerGraph';
+import VisualizerStartScreen from './VisualizerStartScreen';
+import VisualizerToolbar from './VisualizerToolbar';
+import {
+  WorkflowApprovalTemplatesAPI,
+  WorkflowJobTemplateNodesAPI,
+  WorkflowJobTemplatesAPI,
+} from '@api';
 
 const CenteredContent = styled.div`
+  align-items: center;
   display: flex;
   flex-flow: column;
   height: 100%;
-  align-items: center;
   justify-content: center;
 `;
 
-const VisualizerLayout = styled.div`
+const Wrapper = styled.div`
   display: flex;
   flex-flow: column;
   height: 100%;
 `;
 
-const fetchWorkflowNodes = async (templateId, pageNo = 1, nodes = []) => {
-  try {
-    const { data } = await WorkflowJobTemplatesAPI.readNodes(templateId, {
-      page_size: 200,
-      page: pageNo,
-    });
-    if (data.next) {
-      return await fetchWorkflowNodes(
-        templateId,
-        pageNo + 1,
-        nodes.concat(data.results)
-      );
-    }
-    return nodes.concat(data.results);
-  } catch (error) {
-    throw error;
+const fetchWorkflowNodes = async (
+  templateId,
+  pageNo = 1,
+  workflowNodes = []
+) => {
+  const { data } = await WorkflowJobTemplatesAPI.readNodes(templateId, {
+    page_size: 200,
+    page: pageNo,
+  });
+  if (data.next) {
+    return fetchWorkflowNodes(
+      templateId,
+      pageNo + 1,
+      workflowNodes.concat(data.results)
+    );
   }
+  return workflowNodes.concat(data.results);
 };
 
-function Visualizer({ template, i18n }) {
+function Visualizer({ history, template, i18n }) {
```

Comment:
This component is huge. I would look for ways to break it up. Your best bet might be to put a lot of this state into a reducer and add a `useReducer` to manage changes and extract a lot of this state logic into smaller functions.